swap-logic/src/lib.rs

```diff
@@ -0,0 +1,91 @@
+#![cfg_attr(not(test), no_std)]
+
+mod constants;
+mod swap_control;
+
+extern crate alloc;
+
+use alloc::{string::String, vec::Vec};
+
+use crate::constants::methods;
+use contract::{contract_api::runtime, unwrap_or_revert::UnwrapOrRevert};
+use types::{account::PublicKey, ApiError, CLValue, U512};
+
+#[no_mangle]
+pub extern "C" fn delegate() {
+    let method_name: String = runtime::get_arg(0)
+        .unwrap_or_revert_with(ApiError::MissingArgument)
+        .unwrap_or_revert_with(ApiError::InvalidArgument);
+
+    match method_name.as_str() {
+        methods::METHOD_INSERT_KYC_ALLOWANCE_CAP => {
+            let kyc_allowance: U512 = runtime::get_arg(1)
+                .unwrap_or_revert_with(ApiError::MissingArgument)
+                .unwrap_or_revert_with(ApiError::InvalidArgument);
+
+            swap_control::insert_kyc_allowance_cap(kyc_allowance);
+        }
+        methods::METHOD_INSERT_SNAPSHOT_RECORD => {
+            let ver1_address: String = runtime::get_arg(1)
+                .unwrap_or_revert_with(ApiError::MissingArgument)
+                .unwrap_or_revert_with(ApiError::InvalidArgument);
+            let prev_balance: U512 = runtime::get_arg(2)
+                .unwrap_or_revert_with(ApiError::MissingArgument)
+                .unwrap_or_revert_with(ApiError::InvalidArgument);
+
+            swap_control::insert_snapshot(ver1_address, prev_balance);
+        }
+        methods::METHOD_GET_CONTRACT_PURSE => {
```

Comment:
Is there a reason to export this purse?
It seems the contract caller can take over the balance of this purse

Reply:
When swap executing, as users don't know what and where is the contract wallet, I inserted a getter. Please provide the improvement ㅜㅜ

Reply:
Ah, You can narrow the right down like below.

```rust
// swap_control/mod.rs
pub fn get_contract_purse() -> URef {
    let purse = swap_storage::load_contract_wallet();
    URef::new(purse.addr(), AccessRights::READ_ADD) // Users only are able to read and add(amount) the balance of the purse.
}

```

Reply:
Ah, I forgot that users should be able to take the tokens from contract purse.
Check ver1 amount and transfer to user. So, even user can refer contract purse with READ_ADD_WRITE

Reply:
Breach-suspected resolved :)
https://github.com/hdac-io/swap/pull/1/commits/2e8c7578ad68828d8ff46a8e98e10277e8343922